plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx

```diff
@@ -109,3 +119,77 @@ export const addProductCollectionBlockToParentOfPaginationBlock = () => {
 		);
 	}
 };
+
+export const useSetPreviewState = ( {
+	setPreviewState,
+	location,
+	attributes,
+	setAttributes,
+}: {
+	setPreviewState?: SetPreviewState | undefined;
+	location: WooCommerceBlockLocation;
+	attributes: ProductCollectionAttributes;
+	setAttributes: (
+		attributes: Partial< ProductCollectionAttributes >
+	) => void;
+} ) => {
+	const setState = ( newPreviewState: PreviewState ) => {
+		setAttributes( {
+			previewState: {
+				...attributes.previewState,
+				...newPreviewState,
+			},
+		} );
+	};
+
+	// Running setPreviewState function provided by Collection, if it exists.
+	useLayoutEffect( () => {
+		if ( ! setPreviewState ) {
+			return;
+		}
+
+		const cleanup = setPreviewState?.( {
+			setState,
+			location,
+			attributes,
+		} );
+
+		if ( cleanup ) {
+			return cleanup;
+		}
+		// We want this to run only once, adding deps will cause performance issues.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [] );
```

Comment:
It's probably not a big deal, buuut, if the API consumer updated their function how would we know to re-run?

Thought: how dangerous would it be to add `setPreviewState` as a dependency? Worst case, users could trap themselves in perf issues if they don't understand react, but best case, if they want to update the function we would indeed re-run. That'd be my general expectation in react personally.

Reply:
> It's probably not a big deal, buuut, if the API consumer updated their function how would we know to re-run?
> 
> Thought: how dangerous would it be to add `setPreviewState` as a dependency? Worst case, users could trap themselves in perf issues if they don't understand react, but best case, if they want to update the function we would indeed re-run. That'd be my general expectation in react personally. e.g. I am imagining a case where they pass a `useCallback`

IMO it's unlikely that the consumer will change the function itself because the API consumer typically passes this function in the registerProductCollection function like this:

```tsx
registerProductCollection(
	...,
	preview: {
		setPreviewState: ( {
			setState,
			attributes: currentAttributes,
			location,
		}: SetPreviewStateArgs ) => {
			// setPreviewState has access to the current attributes and location.
			console.log( 'setPreviewState' );
			console.log( currentAttributes, location );

			const timeoutID = setTimeout( () => {
				setState( {
					isPreview: false,
					previewMessage: '',
				} );
			}, 5000 );

			return () => clearTimeout( timeoutID );
		},
		initialPreviewState: {
			isPreview: true,
			previewMessage: 'This is in preview mode',
		},
	}
)
```

However, I can add setPreviewState to the dependencies because it shouldn't introduce any performance issues for us. What do you think?

Reply:
> IMO it's unlikely that the consumer will change the function itself because the API consumer typically passes this function in the registerProductCollection function like this:

Yeah probably true, I think we should add it anyway. it's almost no effort and because they'll likely almost never update it then it won't introduce perf issues.